module_auto_update/models/module.py

```diff
@@ -1,70 +1,148 @@
-                    r.display_name)
-
     @api.multi
-    def _store_checksum_installed(self, vals):
-        """Store the right installed checksum, if addon is installed."""
-        if 'checksum_installed' not in vals:
-            try:
-                version = vals["latest_version"]
-            except KeyError:
-                return  # Not [un]installing/updating any addon
-            if version is False:
-                # Uninstalling
-                self.write({'checksum_installed': False})
-            else:
-                # Installing or updating
-                for one in self:
-                    one.checksum_installed = one.checksum_dir
+    def _get_checksum_dir(self):
+        self.ensure_one()
+
+        exclude_patterns = self.env["ir.config_parameter"].get_param(
+            PARAM_EXCLUDE_PATTERNS,
+            DEFAULT_EXCLUDE_PATTERNS,
+        )
+        exclude_patterns = [p.strip() for p in exclude_patterns.split(',')]
+        keep_langs = self.env['res.lang'].search([]).mapped('code')
+
+        module_path = get_module_path(self.name)
+        if module_path and os.path.isdir(module_path):
+            checksum_dir = addon_hash(
+                module_path,
+                exclude_patterns,
+                keep_langs,
+            )
+        else:
+            checksum_dir = False
+
+        return checksum_dir
 
     @api.model
-    def create(self, vals):
-        res = super(Module, self).create(vals)
-        res._store_checksum_installed(vals)
-        return res
+    def _get_saved_checksums(self):
+        Icp = self.env['ir.config_parameter']
+        return json.loads(Icp.get_param(PARAM_INSTALLED_CHECKSUMS, '{}'))
 
-    @api.multi
-    def write(self, vals):
-        res = super(Module, self).write(vals)
-        self._store_checksum_installed(vals)
-        return res
+    @api.model
+    def _save_checksums(self, checksums):
+        Icp = self.env['ir.config_parameter']
+        Icp.set_param(PARAM_INSTALLED_CHECKSUMS, json.dumps(checksums))
+
+    @api.model
+    def _save_installed_checksums(self):
+        checksums = {}
+        installed_modules = self.search([('state', '=', 'installed')])
+        for module in installed_modules:
+            checksums[module.name] = module._get_checksum_dir()
+        self._save_checksums(checksums)
+
+    @api.model
+    def _get_modules_partially_installed(self):
+        return self.search([
+            ('state', 'in', ('to install', 'to remove', 'to upgrade')),
+        ])
+
+    @api.model
+    def _get_modules_with_changed_checksum(self):
+        saved_checksums = self._get_saved_checksums()
+        installed_modules = self.search([('state', '=', 'installed')])
+        return installed_modules.filtered(
+            lambda r: r._get_checksum_dir() != saved_checksums.get(r.name),
+        )
+
+    @api.model
+    def upgrade_changed_checksum(self, overwrite_existing_translations=False):
```

Comment:
Do you think it would be a good idea to add here a different default?: `overwrite_existing_translations=os.environ.get("ODOO_I18N_OVERWRITE", False)`, so we can set this directly through an env variable in automated environments? :thinking:

Reply:
Yes. Not 100% sure about the `ODOO_` prefix, but otherwise :+1: .